src/app/public/modules/text-editor/services/text-editor.service.ts

```diff
@@ -0,0 +1,501 @@
+        const sel = windowEl.getSelection();
+        sel.removeAllRanges();
+        sel.addRange(range);
+      } else {
+        const textRange = editor.createTextRange();
+        textRange.moveToElementText(editor);
+        textRange.collapse(false);
+        textRange.select();
+      }
+    }
+  }
+
+  public getLink(editorId: string): UrlModalResult {
+    let link: UrlModalResult = undefined;
+    const anchorEl = this.getSelectedAnchorTag(editorId);
+    if (anchorEl && anchorEl.href) {
+      link = {
+        target: anchorEl.getAttribute('target') === '_blank' ? UrlTarget.NewWindow : UrlTarget.None,
+        url: anchorEl.href
+      };
+    }
+
+    return link;
+  }
+
+  public getSelectedAnchorTag(editorId: string): HTMLAnchorElement {
+    const selectedEl = this.getCurrentSelectionParentElement(editorId);
+
+    return SkyTextEditorService.getParent(selectedEl, 'a') as HTMLAnchorElement;
+  }
+
+  public getChildSelectedAnchorTags(editorId: string): Element[] {
+    const selectedRange = this.getCurrentSelection(editorId).getRangeAt(0);
+    if (selectedRange.toString().length <= 0) {
+      return [];
+    }
+
+    const parentElement = this.getCurrentSelectionParentElement(editorId);
+    const childElements = parentElement ? Array.from(parentElement.querySelectorAll('a')) : [];
+
+    /* istanbul ignore next */
+    return childElements.filter(element => {
+
+      // IE specific
+      if (!selectedRange.intersectsNode) {
+        if (!element || !element.href) {
+          return false;
+        }
+        const tempRange = document.createRange();
+        tempRange.selectNodeContents(element);
+        return (selectedRange.compareBoundaryPoints(Range.START_TO_START, tempRange) !== -1
+          && selectedRange.compareBoundaryPoints(Range.START_TO_END, tempRange) !== 1) ||
+          (selectedRange.compareBoundaryPoints(Range.END_TO_START, tempRange) !== -1
+          && selectedRange.compareBoundaryPoints(Range.END_TO_END, tempRange) !== 1);
+      }
+
+      // Normal non-IE
+      return !!element && !!element.href && selectedRange.intersectsNode(element);
+    });
+  }
+
+  public getCurrentSelection(id: string): Selection {
+    return this.selectionService.getCurrentSelection(this.getDocumentEl(id));
+  }
+
+  public getCurrentSelectionParentElement(id: string): Element {
+    return this.selectionService.getCurrentSelectionParentElement(this.getDocumentEl(id));
+  }
+
+  public saveSelection(id: string): Range {
+    return this.selectionService.saveSelection(this.getDocumentEl(id), this.getWindowEl(id));
+  }
+
+  public restoreSelection(id: string, range: Range): void {
+    this.selectionService.restoreSelection(this.getDocumentEl(id), range, this.getWindowEl(id));
+  }
+
+  public selectElement(id: string, element: HTMLElement): void {
+    this.selectionService.selectElement(this.getDocumentEl(id), this.getWindowEl(id), element);
+  }
+
+  private getWindowEl(id: string): Window {
```

Comment:
Could we name this `getContentWindowEl`? With the current name I don't think I would recognize that this isn't the native window. I had to look why we were doing this method in one place versus grabbing the native window.

Reply:
Done.